x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -0,0 +1,177 @@
+        }
+
+        @Override
+        public void onPartialReduce(List<SearchShard> shards, TotalHits totalHits, InternalAggregations aggs, int reducePhase) {
+            lastSuccess = shards.size();
+            lastFailures = shardFailures.get();
+            final AsyncSearchResponse newResp = new AsyncSearchResponse(searchId.getEncoded(),
+                new PartialSearchResponse(totalShards, lastSuccess, lastFailures, totalHits, aggs),
+                version.incrementAndGet(),
+                true
+            );
+            response = new Response(newResp, aggs != null);
+        }
+
+        @Override
+        public void onReduce(List<SearchShard> shards, TotalHits totalHits, InternalAggregations aggs) {
+            int failures = shardFailures.get();
+            int ver = (lastSuccess == shards.size() && lastFailures == failures) ? version.get() : version.incrementAndGet();
+            final AsyncSearchResponse newResp = new AsyncSearchResponse(searchId.getEncoded(),
+                new PartialSearchResponse(totalShards, shards.size(), failures, totalHits, aggs),
+                ver,
+                true
+            );
+            response = new Response(newResp, false);
+        }
+
+        @Override
+        public void onResponse(SearchResponse searchResponse) {
+            AsyncSearchResponse newResp = new AsyncSearchResponse(searchId.getEncoded(),
+                searchResponse, version.incrementAndGet(), false);
+            response = new Response(newResp, false);
+        }
+
+        @Override
+        public void onFailure(Exception exc) {
+            AsyncSearchResponse previous = response != null ? response.get(true) : null;
+            response = new Response(new AsyncSearchResponse(searchId.getEncoded(),
+                previous != null ? newPartialResponse(previous, shardFailures.get()) : null,
+                exc != null ? ElasticsearchException.guessRootCauses(exc)[0] : null, version.incrementAndGet(), false), false);
+        }
+
+        private PartialSearchResponse newPartialResponse(AsyncSearchResponse response, int numFailures) {
+            PartialSearchResponse old = response.getPartialResponse();
+            return response.hasPartialResponse() ? new PartialSearchResponse(totalShards, old.getSuccessfulShards(), numFailures,
+                old.getTotalHits(), old.getAggregations()) : null;
+        }
+    }
+
+    private class Response {
+        AsyncSearchResponse internal;
+        boolean needFinalReduce;
+
+        Response(AsyncSearchResponse response, boolean needFinalReduce) {
+            this.internal = response;
+            this.needFinalReduce = needFinalReduce;
+        }
+
+        /**
+         * Ensure that we're performing the final reduce only when users explicitly requested
+         * a response through a {@link GetAsyncSearchAction.Request}.
+         */
+        public synchronized AsyncSearchResponse get(boolean doFinalReduce) {
```

Comment:
why do we need the `synchronized` keyword?

Reply:
I wanted to avoid performing the final reduce multiple times but I simplified the logic and removed the synchronization in https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11 since we expect a single user to access the response.